File: android/app/src/test/java/com/sportsnot/app/widget/WidgetScheduleLayoutTest.kt
package com.sportsnot.app.widget

import com.sportsnot.app.widget.models.League
import com.sportsnot.app.widget.models.WidgetDraftedPlayer
import com.sportsnot.app.widget.models.WidgetGame
import com.sportsnot.app.widget.models.WidgetSnapshot
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class WidgetScheduleLayoutTest {

    @Test
    fun `small pages prioritize live drafted games first`() {
        val snapshot = snapshot(
            games = listOf(
                game(id = 1, startsAt = "2026-04-21T23:00:00Z", state = "FUT"),
                game(id = 2, startsAt = "2026-04-21T22:00:00Z", state = "LIVE"),
                game(id = 3, startsAt = "2026-04-21T21:00:00Z", state = "FUT")
            ),
            players = listOf(
                player(gameId = 1, ownedBy = "12 Bar Hughes", name = "Nick Suzuki"),
                player(gameId = 2, ownedBy = "Connor McJudah", name = "Jake Guentzel")
            )
        )

        val first = WidgetScheduleLayout.pageSections(
            snapshot,
            WidgetHomeSize.SMALL,
            0
        ).single()
        val second = WidgetScheduleLayout.pageSections(
            snapshot,
            WidgetHomeSize.SMALL,
            1
        ).single()
        val third = WidgetScheduleLayout.pageSections(
            snapshot,
            WidgetHomeSize.SMALL,
            2
        ).single()

        assertEquals(2, first.game.id)
        assertEquals(1, second.game.id)
        assertEquals(3, third.game.id)
    }

    @Test
    fun `body text groups assets by fantasy team then nhl team`() {
        val section = WidgetScheduleLayout.pageSections(
            snapshot(
                games = listOf(game(id = 10, startsAt = "2026-04-21T23:00:00Z")),
                players = listOf(
                    player(
                        gameId = 10,
                        ownedBy = "12 Bar Hughes",
                        teamAbbrev = "MTL",
                        name = "Cole Caufield",
                        fantasyPoints = 5.0
                    ),
                    player(
                        gameId = 10,
                        ownedBy = "12 Bar Hughes",
                        teamAbbrev = "MTL",
                        name = "Nick Suzuki",
                        fantasyPoints = 7.0
                    ),
                    player(
                        gameId = 10,
                        ownedBy = "12 Bar Hughes",
                        teamAbbrev = "TBL",
                        name = "Nikita Kucherov",
                        fantasyPoints = 9.0
                    ),
                    player(
                        gameId = 10,
                        ownedBy = "Connor McJudah",
                        teamAbbrev = "TBL",
                        name = "Jake Guentzel",
                        fantasyPoints = 4.0
                    )
                )
            ),
            WidgetHomeSize.LARGE,
            0
        ).single()

        val body = WidgetScheduleLayout.bodyText(section, WidgetHomeSize.LARGE)

        assertTrue(body!!.contains("12 Bar Hughes"))
        assertTrue(body.contains("- MTL: Nick Suzuki, Cole Caufield"))
        assertTrue(body.contains("- TBL: Nikita Kucherov"))
        assertTrue(body.contains("Connor McJudah"))
    }

    @Test
    fun `medium pages chunk by games`() {
        val snapshot = snapshot(
            games = listOf(
                game(id = 1, startsAt = "2026-04-21T20:00:00Z"),
                game(id = 2, startsAt = "2026-04-21T21:00:00Z"),
                game(id = 3, startsAt = "2026-04-21T22:00:00Z")
            ),
            players = emptyList()
        )

        assertEquals(
            2,
            WidgetScheduleLayout.totalPages(snapshot, WidgetHomeSize.MEDIUM)
        )
        assertEquals(
            listOf(1, 2),
            WidgetScheduleLayout.pageSections(
                snapshot,
                WidgetHomeSize.MEDIUM,
                0
            ).map { it.game.id }
        )
        assertEquals(
            listOf(3),
            WidgetScheduleLayout.pageSections(
                snapshot,
                WidgetHomeSize.MEDIUM,
                1
            ).map { it.game.id }
        )
    }

    private fun snapshot(
        games: List<WidgetGame>,
        players: List<WidgetDraftedPlayer>
    ) = WidgetSnapshot(
        league = League(
            id = "league-1",
            name = "Test League",
            shareCode = "ABC123",
            currentRound = 1,
            status = "active"
        ),
        date = "2026-04-21",
        generatedAt = "2026-04-21T18:00:00Z",
        games = games,
        players = players
    )

    private fun game(
        id: Int,
        startsAt: String,
        state: String = "FUT"
    ) = WidgetGame(
        id = id,
        startsAt = startsAt,
        state = state,
        homeTeamId = id * 10,
        homeTeamAbbrev = "T$id",
        homeTeamName = "Home $id",
        homeScore = 0,
        awayTeamId = id * 10 + 1,
        awayTeamAbbrev = "A$id",
        awayTeamName = "Away $id",
        awayScore = 0,
        period = if (state == "LIVE") 2 else null,
        timeRemaining = if (state == "LIVE") "14:07" else null,
        hasDraftedPlayers = true
    )

    private fun player(
        gameId: Int,
        ownedBy: String,
        teamAbbrev: String = "MTL",
        name: String,
        fantasyPoints: Double = 3.0
    ) = WidgetDraftedPlayer(
        playerId = null,
        teamId = null,
        name = name,
        teamAbbrev = teamAbbrev,
        position = "F",
        gameId = gameId,
        fantasyPoints = fantasyPoints,
        ownedByTeamName = ownedBy
    )
}

File: packages/web/src/app/routes/leagues/widgetScheduleLayout.test.ts
import { describe, expect, it } from '@rstest/core';
import type { WidgetSnapshot } from '@sportsnot/widget-api';
import {
  buildWidgetGameCards,
  formatWidgetGameHeader,
  formatWidgetTeamLines,
} from './widgetScheduleLayout';

function makeSnapshot(): WidgetSnapshot {
  return {
    league: {
      id: 'league-1',
      name: 'Playoff League',
      shareCode: 'SHARE',
      currentRound: 2,
      status: 'active',
    },
    date: '2026-04-23',
    generatedAt: '2026-04-23T12:00:00Z',
    games: [
      {
        id: 20,
        startsAt: '2026-04-24T02:00:00Z',
        state: 'LIVE',
        homeTeamId: 52,
        homeTeamAbbrev: 'WPG',
        homeTeamName: 'Jets',
        homeScore: 3,
        awayTeamId: 25,
        awayTeamAbbrev: 'DAL',
        awayTeamName: 'Stars',
        awayScore: 2,
        period: 2,
        timeRemaining: '11:02',
        hasDraftedPlayers: true,
      },
      {
        id: 10,
        startsAt: '2026-04-23T23:00:00Z',
        state: 'OFF',
        homeTeamId: 12,
        homeTeamAbbrev: 'CAR',
        homeTeamName: 'Hurricanes',
        homeScore: 4,
        awayTeamId: 13,
        awayTeamAbbrev: 'FLA',
        awayTeamName: 'Panthers',
        awayScore: 1,
        period: 3,
        timeRemaining: null,
        hasDraftedPlayers: false,
      },
    ],
    players: [
      {
        playerId: 1,
        teamId: null,
        name: 'Roope Hintz',
        teamAbbrev: 'DAL',
        position: 'F',
        gameId: 20,
        fantasyPoints: 4,
        dailyFantasyPoints: 1,
        ownedByTeamName: 'Alpha',
      },
      {
        playerId: 2,
        teamId: null,
        name: 'Jason Robertson',
        teamAbbrev: 'DAL',
        position: 'F',
        gameId: 20,
        fantasyPoints: 1,
        dailyFantasyPoints: 0,
        ownedByTeamName: 'Alpha',
      },
      {
        playerId: 3,
        teamId: null,
        name: 'Connor Hellebuyck',
        teamAbbrev: 'WPG',
        position: 'G',
        gameId: 20,
        fantasyPoints: 3,
        dailyFantasyPoints: 2,
        ownedByTeamName: 'Bravo',
      },
    ],
  };
}

describe('widgetScheduleLayout', () => {
  it('buildWidgetGameCards sorts games chronologically', () => {
    const cards = buildWidgetGameCards(makeSnapshot());

    expect(cards.map((card) => card.game.id)).toEqual([10, 20]);
  });

  it('buildWidgetGameCards groups fantasy teams by total points', () => {
    const cards = buildWidgetGameCards(makeSnapshot());

    expect(cards[1].fantasyTeams.map((team) => team.name)).toEqual([
      'Alpha',
      'Bravo',
    ]);
    expect(formatWidgetTeamLines(cards[1].fantasyTeams[0])).toEqual([
      '- DAL: Roope Hintz 4 +1, Jason Robertson 1 +0',
    ]);
  });

  it('formatWidgetGameHeader matches live and final widget text', () => {
    const [finalGame, liveGame] = makeSnapshot().games.sort(
      (lhs, rhs) => lhs.id - rhs.id
    );

    expect(formatWidgetGameHeader(finalGame)).toBe('FLA @ CAR - 1-4 F');
    expect(formatWidgetGameHeader(liveGame)).toBe('DAL @ WPG - 2-3 P2 11:02');
  });
});
